added signal "button_pressed" to ButtonArray base class

brunosxs opened the pull request:
ButtonArray has a "button_selected" signal, but not a "button_pressed" signal, which makes code in gdscript more complex to handle situations when you want to get the button that the user really clicked, instead of just "selected". The buttons that inherit from BaseButton all have the "pressed" signal, which made this puzzling for me. So, here is said fix.

I will also attach an example project with everything setted to test this new signal with(as of now) the current master branch.
[bArrayButton_signal_pressed.zip](https://github.com/godotengine/godot/files/1071425/bArrayButton_signal_pressed.zip)

brunosxs committed:
added signal "button_pressed" to ButtonArray base class

ButtonArray has a "button_selected" signal, but not a "button_pressed" signal, which makes code in gdscript more complex to handle situations when you want to get the button that the user really clicked, instead of just "selected". The buttons that inherit from BaseButton all have the "pressed" signal, which made this puzzling for me. So, here is said fix.

volzhs commented:
There are plenty of unnecessary changes with your commit.
make sure to use tab instead of spaces with your IDE.
and this will also help.
https://github.com/godotengine/godot/tree/master/misc/hooks

brunosxs commented:
Thanks for the feedback, I fixed the formatting now. Didn't knew QTCreator used spaces instead of tabs.
I've also added the hooks to the .git hooks folder and it reported no problems.

volzhs commented on scene/gui/button_array.cpp:
```diff
@@ -309,6 +309,14 @@ void ButtonArray::_gui_input(const Ref<InputEvent> &p_event) {
 	}
 
+	if (
+			p_event->is_action("ui_accept") && p_event->is_pressed() && selected < (buttons.size())) {
```
there is unnecessary new line here.

volzhs commented on scene/gui/button_array.cpp:
```diff
@@ -320,6 +328,21 @@ void ButtonArray::_gui_input(const Ref<InputEvent> &p_event) {
+		for (int i = 0; i < buttons.size(); i++) {
+
+			if (ofs >= buttons[i]._pos_cache && ofs < buttons[i]._pos_cache + buttons[i]._size_cache) {
+
```
would you remove 4 new lines on up here?

brunosxs replied:
if I remove those, the signal won't be able to fire when the mouse triggers the event.(or is there a better implementation to this?)
I am fairly new to the c++ workflow, so excuse any stupid mistakes..

volzhs replied:
@brunosxs I mean "empty new lines"...

brunosxs commented:
I think it is done now.

Zylann commented:
I don't get the difference between `button_selected` and `button_pressed`... aren't they the same in a ButtonArray?

brunosxs commented:
You can "select" a button when you grab its focus... you can do that with a keyboard by moving the arrow keys towards a button or combining that with tab for an example.
But pressed is when you actually hit it with the mouse or hit the ui_accept... with the button selected.
Those are different behaviors and it would be for the better to separate them.
They are separated in the BaseButton class, you have a signal for selected and another for pressed... (I always found the names "selected" and "pressed" confusing haha)

akien-mga commented:
@brunosxs We reviewed some PRs yesterday on IRC, and when it came to this one, though we agreed on the API addition, we finally decided to remove ButtonArray completely as it serves the same purpose as Buttons + a V/HBoxContainer, albeit with a more limited API.

See http://godot.eska.me/irc-logs/devel/2017-06-14.log from 23:16 to 23:28.

I think your PR might still be relevant at least in part for the 2.1 branch though.

Zylann commented:
@akien-mga how should we do mutually exclusive buttons then? I found ButtonArrays to be handy at this.

bojidar-bg commented:
@Zylann I think button arrays are getting replaced by the ButtonGroup resource (but I can't tell for sure).

brunosxs commented:
I read the log for the IRC chat, thanks for the consideration haha. I indeed learned c++ just to do this, but it is a skill that is fun to have, I'll keep improving on it as it will help me in other projects. Besides godot's ecossystem helped me evolve a lot as a programmer.

Well, if you guys say so, no problem. I thought at the time ButtonArrays were useful(still do) because they're easier to setup, you just need to connect a single time to handle an unlimited number of buttons... it was just lacking in the respect that "button_pressed" acted the way it did, and there was no "pressed" signal.

In my case, I needed to first see if the items were selected with the signal "button_selected"
https://github.com/brunosxs/SMRT-Godot/blob/9f350ce9038c1ac2e22052787c1d82c65c75e2a7/addons/SMRT/dialog.gd#L363-L366

and them do some more checkings in input, adding things like keeping a variable dedicated to the answer selected so I could move the index around the functions of the script.
https://github.com/brunosxs/SMRT-Godot/blob/9f350ce9038c1ac2e22052787c1d82c65c75e2a7/addons/SMRT/dialog.gd#L368-L376
To work around the fact that a signal for "pressed didn't exist"

And this would be much more simplified(less lines) if I could simply yield for a button_pressed in a ButtonArray. They naturally have their index, something that doesn't exist in buttons as you pass the reference to the button itself(of course I could grab the index of it relative to the parent node). But the simpler setup of ButtonArrays is more charmingly easier for me.

Looking at the documentation now, ButtonGroups also suffer from the same problem: they only have a "selected" signal. It would still be useful to create/separate the behaviors "selected" from pressed in different signals.

Well, that's basically my specific case.

karroffel commented:
@brunosxs do you want to port this to 2.1 though?

karroffel commented:
If you want you can open a new PR for 2.1, but for 3.0 this is no longer relevant due to the removal of ButtonArray.
